framework/src/main/java/org/checkerframework/framework/type/AbstractViewpointAdapter.java

```diff
@@ -0,0 +1,492 @@
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedExecutableType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedNullType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedPrimitiveType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedTypeVariable;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedWildcardType;
+import org.checkerframework.javacutil.ElementUtils;
+import org.checkerframework.javacutil.ErrorReporter;
+import org.checkerframework.javacutil.Pair;
+
+/**
+* Abstract utility class for performing viewpoint adaptation.
+*
+* This class contains the common logic for extracting and inserting viewpoint adapted annotations
+* into the coresponding types for member/field access, constructor and method invocations, and
+* type parameter bound instantiations.
+*
+* Subclasses implement the computation of the precise viewpoint adapted type given a receiver type
+* and a declared type, and implement how to extract the modifier given an ATM.
+*/
+public abstract class AbstractViewpointAdapter implements ViewpointAdapter {
+
+    // This prevents calling combineTypeWithType on type variable if it is an upper bound
+    // of another type variable. We only viewpoint adapt type variable that is not upper-bound.
+    private boolean isTypeVarExtends = false;
+
+    protected final AnnotatedTypeFactory atypeFactory;
+
+    public AbstractViewpointAdapter(final AnnotatedTypeFactory atypeFactory) {
+        this.atypeFactory = atypeFactory;
+    }
+
+    @Override
+    public void viewpointAdaptMember(
+            AnnotatedTypeMirror receiverType,
+            Element memberElement,
+            AnnotatedTypeMirror memberType) {
+        if (!shouldAdaptMember(memberType, memberElement)) {
+            return;
+        }
+
+        AnnotatedTypeMirror decltype = atypeFactory.getAnnotatedType(memberElement);
+        AnnotatedTypeMirror combinedType = combineTypeWithType(receiverType, decltype);
+        memberType.replaceAnnotations(combinedType.getAnnotations());
+        if (memberType.getKind() == TypeKind.DECLARED
+                && combinedType.getKind() == TypeKind.DECLARED) {
+            AnnotatedDeclaredType adtType = (AnnotatedDeclaredType) memberType;
+            AnnotatedDeclaredType adtCombinedType = (AnnotatedDeclaredType) combinedType;
+            adtType.setTypeArguments(adtCombinedType.getTypeArguments());
+        } else if (memberType.getKind() == TypeKind.ARRAY
+                && combinedType.getKind() == TypeKind.ARRAY) {
+            AnnotatedArrayType aatType = (AnnotatedArrayType) memberType;
+            AnnotatedArrayType aatCombinedType = (AnnotatedArrayType) combinedType;
+            aatType.setComponentType(aatCombinedType.getComponentType());
+        }
+    }
+
+    protected boolean shouldAdaptMember(AnnotatedTypeMirror type, Element element) {
```

Comment:
in the implementation here, the ATM isn't used at all in the method body. Can you add javadoc explaining why this parameter is needed or how it should be used?